### ISSUE TITLE
Change Request: Backwards compatibility utility for rules

### ISSUE THREAD

**Issue by nzakas** [eslint/eslint#18398]
### ESLint version

HEAD

### What problem do you want to solve?

After talking with folks online, it seems that one of the big blockers for people upgrading to ESLint v9.x is that there are still a fair number of plugins that haven't made the changes we outlined in the [blog post](https://eslint.org/blog/2023/09/preparing-custom-rules-eslint-v9/). This leaves users stuck because they can't migrate to v9.x when their config doesn't work.

### What do you think is the correct solution?

I propose creating a new package, `@eslint/backcompat`, that contains functions that wrap existing rules to create a `context` object that looks like the pre-9.x object. This package would export three functions:

1. `fixupRule(rule)` - to be applied directly to rule objects
2. `fixupPluginRules(plugin)` - given a plugin, returns an object where all of the rules are wrapped with `fixupRule()`
3. `fixConfigRules(configs)` - given an array of config objects, wraps all of the plugin definitions using `fixupPluginRules()`.

In most cases people would probably use the package like this:

```js
import { fixupConfigRules } from "@eslint/backcompat";

export default fixupConfigRules([
   // ... their config
]);
```

### Participation

- [X] I am willing to submit a pull request for this change.

**Comment by mdjermanovic** [comment]
Sounds good to me, though it might be tricky for the wrapper rule to determine the currently traversed node, which is needed when calling some of the new methods (e.g., `SourceCode#getScope(node)`).

**Comment by nzakas** [comment]
I actually already have a prototype of this working. :+1:

**Comment by aladdin-add** [comment]
SGTM! 👍 

Re the naming of this package: there may be other versions needed in the future, considering`@eslint/backcompat-v9`?
